### ISSUE TITLE
Confusing int vs long with regards to FileOperations2.advanceBytes

### ISSUE THREAD

**Issue by stevenschlansker** [sunnygleason/g414-hash#10]
Hi,

It seems that the size of HashFile2s is limited by Integer.MAX_VALUE --

```
/**
 * Advances the file pointer by <code>count</code> bytes, throwing an
 * exception if the postion has exhausted a long (hopefully not likely).
 */
private static long advanceBytes(long pos, long count, boolean isLongPos)
        throws IOException {
    long newpos = pos + count;
    if (newpos < count || newpos > Integer.MAX_VALUE)
        throw new IOException("HashFile is too big.");
    return newpos;
}
```

All of the arguments are lovingly created as longs, but if the position ever grows past Integer.MAX_VALUE (which is only ~2GB) then a "HashFile too big" exception is thrown.

**Comment by sunnygleason** [comment]
Can you test with Long.MAX_VALUE ? Looking into that - I'm wondering if it's a simple typo, or an intentional limitation due to a limit on file.seek or bytebuffer size

**Comment by stevenschlansker** [comment]
Will test and report back.

That said, the check makes no sense as a long could never be > Long.MAX_VALUE...

**Comment by stevenschlansker** [comment]
Hm, quick update...  Been waiting about 24 hours for my HashFile2Builder to finish() on a 20GB hash file with the patch applied.  Since there's no progress information, I have no idea what / if it's doing :-(

Might hack up some logging so that there's some visibility into this.  Do you have a preferred logging library, or is this not a patch you'd like?

**Comment by sunnygleason** [comment]
Hi Steven - one thing you might try before going too far is seeing if HashFile works for you (pretty much the same interface).

HashFile is 64-bit compatible. I just started HashFile2 so I could add some features, which probably also introduced the bug you're hitting.

Would you be able to give com.g414.hash.file.HashFile a try?

**Comment by stevenschlansker** [comment]
Sure, sounds good.  I picked HashFile2 because 2 sounds newer, no other reason.  Will report back again.

**Comment by sunnygleason** [comment]
In the meantime, I'll try to replicate your issue. What are the specs of your hashfile data? total kv pairs, # of distinct keys, # of values per key, byte sizes of keys and values? Thanks in advance for any info you can provide--

**Comment by stevenschlansker** [comment]
13M keys.  I have a few files, and the key distribution is either unique (every key unique, 1 value) or pretty much the opposite (13M keys but only 10 or so distinct values, each with ~1M values).

Keys range from 5-16 bytes and values range from 16bytes-2KBytes

**Comment by stevenschlansker** [comment]
Aha, it seems to be the one with very poor key distribution that takes the longest to write out.  All the other files finish reasonably quickly, but that one is taking an order of magnitude longer to create  (The others have flushed out all 256 radix buckets but this one is still on 15).  (Estimates, still running).  That said it could be that once it hits the data the rest complete immediately.

**Comment by stevenschlansker** [comment]
Okay, so by switching to HashFile and removing the degenerate index, everything seems to work great!  HashFile2 still has questions about the Integer.MAX_INT limitation, and neither coped well with the degenerate case.  Not sure if you care to consider the latter a problem, as this is very much an edge case and might not be worth the effort to fix.

Thanks again!

**Comment by sunnygleason** [comment]
Awesome, so glad that works! I'll investigate the issues with hashfile2. I think the degenerate case might be insurmountable given the data structure, but I'll ponder it some more!

-Sunny

**Comment by sunnygleason** [comment]
@stevenschlansker, I went through and added a bunch more extensive tests to HashFile2 and shook out a _lot_ of bugs. Feel free to give it a try if you like! I'm just running exhaustive tests and will create a new maven release in the next day or 2.

I think that the degenerate case is pretty tough to beat. The only thing I can think of to help spread keys throughout the buckets is to append a random small integer to the key before inserting, and when querying, query all keys.

Hope this helps! -Sunny

**Comment by stevenschlansker** [comment]
Awesome!  I actually just got things working with HashFile and have pretty much used up the amount of time I can spend on this at work, but I will definitely give it a try next time I revisit our usage and report back.  Thanks again for the awesome library and support.

Understandable that the degenerate case sucks, so I actually just dropped that index from our data.  Hopefully nobody will notice o:)
